metadeploy/api/views.py

```diff
@@ -176,51 +191,151 @@ def filter_get_one(self, qs):
         serializer = PreflightResultSerializer(instance=preflight)
         return Response(serializer.data)
 
     def preflight_post(self, request):
+        scratch_org_id = request.session.get("scratch_org_id", None)
         plan = get_object_or_404(Plan.objects, id=self.kwargs["pk"])
         is_visible_to = plan.is_visible_to(
             request.user
         ) and plan.version.product.is_visible_to(request.user)
-        if not is_visible_to:
+        user_kwargs = {
+            "user": request.user if request.user.is_authenticated else None,
+            "uuid": scratch_org_id,
+        }
+        if not is_visible_to or not any(user_kwargs.values()):
             return Response("", status=status.HTTP_403_FORBIDDEN)
+
+        forced_user_kwargs = {}
+        try:
+            scratch_org_job = ScratchOrgJob.objects.get(uuid=scratch_org_id)
+            config = scratch_org_job.config
+            org_kwargs = {
+                "organization_url": config["instance_url"],
+                "org_id": config["org_id"],
+            }
+            # Are these values in the config? In jobs.py they're a bit
+            # different.
+            forced_user_kwargs["forced_user"] = FakeUser(
+                token=(config["access_token"], config["refresh_token"]),
+            )
+        except (ScratchOrgJob.DoesNotExist, ScratchOrgJob.MultipleObjectsReturned):
+            org_kwargs = {
+                "organization_url": request.user.instance_url,
+                "org_id": request.user.org_id,
+            }
+
         preflight_result = PreflightResult.objects.create(
-            user=request.user,
             plan=plan,
-            organization_url=request.user.instance_url,
-            org_id=request.user.org_id,
+            **org_kwargs,
+            **user_kwargs,
         )
-        preflight_job.delay(preflight_result.pk)
+        preflight_job.delay(preflight_result.pk, **forced_user_kwargs)
         serializer = PreflightResultSerializer(instance=preflight_result)
         return Response(serializer.data, status=status.HTTP_201_CREATED)
 
-    @action(detail=True, methods=["post", "get"])
+    @action(detail=True, methods=["post", "get"], permission_classes=(AllowAny,))
     def preflight(self, request, pk=None):
         if request.method == "GET":
             return self.preflight_get(request)
         if request.method == "POST":
             return self.preflight_post(request)
 
+    def scratch_org_get(self, request):
+        scratch_org_id = request.session.get("scratch_org_id", None)
+        plan = get_object_or_404(Plan.objects, id=self.kwargs["pk"])
+        scratch_org = ScratchOrgJob.objects.filter(
+            uuid=scratch_org_id, plan=plan
+        ).first()
+        if scratch_org is None:
+            return Response("", status=status.HTTP_404_NOT_FOUND)
+        serializer = ScratchOrgJobSerializer(instance=scratch_org)
+        return Response(serializer.data)
```

Comment:
@wlonk I went ahead and threw this in; please review/cleanup and backfill tests.